### PR TITLE
Add PHP compatibility checks to PRs

### LINKED ISSUE #3326: Improve PHP compatibility checking [Investigation]
### Description
WooCommerce Payments 3.2.0 shipped with [a bug](https://github.com/Automattic/woocommerce-payments/issues/3244) that caused fatal errors on stores running on PHP version 7.2 or lower. This required a subsequent patch release.

WCPay 3.2.0 was the first release to include `woocommerce-subscriptions-core` as a composer dependency in WooCommerce Payments. The bug that caused the fatal error was tracked down to a syntax issue (trailing commas in function calls) in `woocommerce-subscription-core`, it was fixed in this PR - https://github.com/Automattic/woocommerce-subscriptions-core/pull/23.

`woocommerce-payments` performs PHP [compatibility tests](https://github.com/Automattic/woocommerce-payments/blob/develop/.github/workflows/compatibility.yml) on pull requests but did not detect this issue.

As far as I can tell, the existing compatibility checks simply run the PHP unit tests across a combination of WC, WP and PHP versions. It also seems the majority of the tests are run against PHP 7.4 (see the image below).

Screen Shot 2021-11-09 at 11 24 21 am

This is problematic as the existing unit tests do not cover all possible code execution paths in the codebase. I think it'd be better to use a tool that explicitly checks PHP compatibility across the entire codebase.

There might be more thorough ways to test PHP compatibility:
- ~Consider using [PHPCompatibility](https://github.com/PHPCompatibility/PHPCompatibility) for PHP compatibility checking.~ **Update:** It appears we're already using [woocommerce/woocommerce-sniffs](https://github.com/Automattic/woocommerce-payments/blob/0a23b79/composer.json#L26) which uses [PHPCompatibilityWP](https://github.com/PHPCompatibility/PHPCompatibilityWP) which uses PHPCompatibility.
- Consider running PHP compatibility checks against a production package. This provides greater confidence that the final package being shipped is compatible with various PHP versions (including vendor files).

### Acceptance criteria

* Decide on the best approach to improve PHP compatibility checking in `woocommerce-payments` (see the proposed solutions above). These checks _must_ ensure that any shipped vendor code (e.g. `woocommerce-subscriptions-core`) is compatible with supported PHP versions.
* Create a new GitHub ticket detailing this approach.
* Close this GitHub ticket.

### Additional context
Incident report - paJDYF-2OJ-p2

### PR DESCRIPTION
**Pull request by RadoslavGeorgiev**
Fixes #3326

#### Changes proposed in this Pull Request

Add a new workflow, which runs the `PHPCompatibility` rules on both local files, and production dependencies.

We already have the `woocommerce/woocommerce-sniffs` dependency, which includes everything necessary to check for PHP compatibility, but until now it was only used for WCPay-specific code.

This PR includes 3 changes:

1. Create a new `phpcs-compat.xml.dist` file, based on the existing `phpcs.xml.dist`. The new file removes all other rules, except for `PHPCompatibility`, which we need.
2. Create a new `phpcs-compat.sh` bash script, which:
    - Does a standard Composer install, which will allow access to the `phpcs` command.
    - Performs an additional Composer install, but into a separate `vendor-dist` directory, and with the `--no-dev` flag, in order to only include production dependencies. The new directory is not ignored in the configuration, so it will be cheked.
    - Actually runs `phpcs` with the new configuration.
3. An additional workflow, which uses a minimal configuration in order to run the new script. It runs on PHP 7.4, but this does not matter, as the code sniffer will catch issues based on the `minimum_supported_wp_version` and `testVersion` values. Those are set to `7.0`, as the plugin requires this version.

Now we'll be sure what we're not introducing PHP incompatibility through WCPay, and its dependencies.

**Note:** The workflow currently fails, because there are a couple of errors. If we agree to this approach, we should ask for them to be fixed before merging this PR.

#### Testing instructions

Check how the workflow currently runs, and its output.

Additionally, you can do a local test:

1. Comment the two `rm -rf vendor-dist` lines in `bin/phpcs-compat.sh`.
2. Run `./bin/phpcs-compat.sh`. This will output some errors, ignore them for now.
3. Navigate to any PHP file within the new `vendor-dist` and introduce a PHP compatibility error.
4. Run `./bin/phpcs-compat.sh` again. There should be an error, related to the change.

Here's how this worked out for me when I added a comma at the end of a function's parameter list:

Screenshot 2021-11-09 at 16 17 26

-------------------

- [x] No changelog entry required
- [x] Covered with tests (it is basically tests)

**Before merge**

- [ ] Make sure that all existing PHP errors/warnings are either ignored or fixed.

### CODE REVIEW

**Comment by dmvrtx** [comment]
Wouldn't it be better to add this PHPCodeSniffer check as a local pre-commit hook? We already run it with main rule set, so one more run shouldn't present a big problem, but could catch errors earlier in the process.

File: phpcs-compat.xml.dist
Comment (aprea) [inline phpcs-compat.xml.dist:18]: This should be the WordPress version, not the PHP version.

Ref: https://github.com/WordPress/WordPress-Coding-Standards/wiki/Customizable-sniff-properties#setting-minimum-supported-wp-version-for-all-sniffs-in-one-go-wpcs-0140

File: .gitignore
Comment (aprea) [inline .gitignore:60]: Can we add this as an exclude pattern in `phpcs.xml.dist`?

**Comment by aprea** [comment]
This is a neat approach, nice one Rado! I left a few comments but otherwise, this looks like it'll work 🎉

File: phpcs-compat.xml.dist
Comment (RadoslavGeorgiev) [reply phpcs-compat.xml.dist]: Thank you for noticing! Somehow I missed the `wp_` prefix, but now it's changed to 5.6, the same as the `Requires WP` plugin header.

File: .gitignore
Comment (RadoslavGeorgiev) [reply .gitignore:60]: Done in https://github.com/Automattic/woocommerce-payments/pull/3330/commits/5b42ac353ebadb4d64c777b938659f70a6570740 ✅

**Comment by RadoslavGeorgiev** [comment]
All feedback so far has been addressed, but I'll wait for another review round before I open PRs for the existing issues.

File: .github/workflows/compatibility.yml
Comment (deepakpathania) [inline .github/workflows/compatibility.yml:43]: Nitpick - extra new line seems unrelated to changes.

File: .github/workflows/php-compatibility.yml
Comment (deepakpathania) [inline .github/workflows/php-compatibility.yml:4]: Do we want to specify target branches for this?

File: .github/workflows/php-compatibility.yml
Comment (deepakpathania) [inline .github/workflows/php-compatibility.yml:13]: Unrelated but does it make sense to fork this and keep the things we need?

**Comment by deepakpathania** [comment]
Verified that the workflow works as intended.
On local, initially saw these errors which are known:

Screenshot 2021-11-30 at 6 17 14 PM

Then introduced an error and ran the workflow again and could see that it was shown in the command output.

Screenshot 2021-11-30 at 6 19 38 PM


>  If we agree to this approach, we should ask for them to be fixed before merging this PR.

@RadoslavGeorgiev I think we can go ahead and fix the existing issues now in this PR.

File: .github/workflows/compatibility.yml
Comment (RadoslavGeorgiev) [reply .github/workflows/compatibility.yml]: That's correct. The line was already in place, but it seems like my editor decided to add an extra space for some reason, which is now removed.

File: .github/workflows/php-compatibility.yml
Comment (RadoslavGeorgiev) [reply .github/workflows/php-compatibility.yml:4]: I don't think so. The point is to execute the script on every pull request to make authors aware of risky PHP code, and this line will run it on the PR branch. Do you think we need to?

File: .github/workflows/php-compatibility.yml
Comment (RadoslavGeorgiev) [reply .github/workflows/php-compatibility.yml:13]: It's just a tool that we use to setup PHP, and it works. We could fork it and strip the stuff we don't need, but I don't see any meaningful reason to do so. Did you have something in mind?

**Comment by RadoslavGeorgiev** [comment]
Thank you for the review, @deepakpathania!

> I think we can go ahead and fix the existing issues now in this PR.

Those issues are in other repositories/packages :D

File: .github/workflows/php-compatibility.yml
Comment (deepakpathania) [reply .github/workflows/php-compatibility.yml:4]: My thought was to not keep the action workers busy on temporary branches we might push but since the workflow doesn't take that long I think it is okay to run it on every branch.

File: .github/workflows/php-compatibility.yml
Comment (deepakpathania) [reply .github/workflows/php-compatibility.yml:13]: Not really, the repository name just stood out since it was out of org but I guess it is fine to have it here.

**Review by deepakpathania — APPROVED** [review]
LGTM 🚢

**Comment by RadoslavGeorgiev** [comment]
Turned out that the warnings can be ignored for all reports, so I did that in our PHPCS compatibility configuration.

**Pull request merged**
